pygeo/constraints/thicknessConstraint.py

```diff
@@ -88,6 +88,115 @@ def writeTecplot(self, handle):
             handle.write("%d %d\n" % (2 * i + 1, 2 * i + 2))
 
 
+class ProjectedThicknessConstraint(GeometricConstraint):
+    """
+    DVConstraints representation of a set of projected thickness
+    constraints. One of these objects is created each time a
+    addThicknessConstraints2D or addThicknessConstraints1D call is
+    made. The user should not have to deal with this class directly.
+
+    This is different from ThicknessConstraints becuase it measures the projected
+    thickness along the orginal direction of the constraint.
+    """
+
+    def __init__(self, name, coords, lower, upper, scaled, scale, DVGeo, addToPyOpt, compNames):
+        super().__init__(name, len(coords) // 2, lower, upper, scale, DVGeo, addToPyOpt)
+
+        self.coords = coords
+        self.scaled = scaled
+
+        # First thing we can do is embed the coordinates into DVGeo
+        # with the name provided:
+        self.DVGeo.addPointSet(self.coords, self.name, compNames=compNames)
+
+        # Now get the reference lengths and directions
+        self.D0 = np.zeros(self.nCon)
+        self.dir_vec = np.zeros((self.nCon, 3))
+        for i in range(self.nCon):
+            vec = self.coords[2 * i] - self.coords[2 * i + 1]
+            self.D0[i] = geo_utils.norm.euclideanNorm(vec)
+            self.dir_vec[i] = vec / self.D0[i]
+
+    def evalFunctions(self, funcs, config):
+        """
+        Evaluate the functions this object has and place in the funcs dictionary
+
+        Parameters
+        ----------
+        funcs : dict
+            Dictionary to place function values
+        """
+        # Pull out the most recent set of coordinates:
+        self.coords = self.DVGeo.update(self.name, config=config)
+        D = np.zeros(self.nCon)
+        for i in range(self.nCon):
+            vec = self.coords[2 * i] - self.coords[2 * i + 1]
+
+            # take the dot product with the direction vector
+            D[i] = vec[0] * self.dir_vec[i, 0] + vec[1] * self.dir_vec[i, 1] + vec[2] * self.dir_vec[i, 2]
+
+            if self.scaled:
+                D[i] /= self.D0[i]
+
+        funcs[self.name] = D
+
+    def evalFunctionsSens(self, funcsSens, config):
+        """
+        Evaluate the sensitivity of the functions this object has and
+        place in the funcsSens dictionary
+
+        Parameters
+        ----------
+        funcsSens : dict
+            Dictionary to place function values
+        """
+
+        nDV = self.DVGeo.getNDV()
+        if nDV > 0:
+            dTdPt = np.zeros((self.nCon, self.coords.shape[0], self.coords.shape[1]))
+            for i in range(self.nCon):
+                D_b = 1.0
+
+                if self.scaled:
+                    D_b /= self.D0[i]
+
+                # d(dot(vec,n)/dvec = n
```

Comment:
Do you mind making this derivative more explicit? There's a bracket missing but it looks like replacing `eDist_b` is the main step here.

Reply:
I have added more comments let me know what you think